chain/src/main/scala/org/bitcoins/chain/blockchain/Blockchain.scala

```diff
@@ -53,103 +54,153 @@ case class Blockchain(headers: Vector[BlockHeaderDb])
+
+          tipResult match {
+            case success: TipUpdateResult.Success =>
+              logger.debug(
+                s"Successfully verified=${success.header.hashBE.hex}, connecting to chain")
+              val connectionIdx = blockchain.length - prevHeaderIdx
+
+              val oldChain =
+                blockchain.takeRight(connectionIdx)
+              val newChain =
+                Blockchain.fromHeaders(success.headerDb +: oldChain)
+
+              if (connectionIdx != blockchain.length) {
+                //means we have a reorg, since we aren't connecting to latest tip
+                ConnectTipResult.Reorg(success, newChain)
+              } else {
+                //we just extended the latest tip
+                ConnectTipResult.ExtendChain(success, newChain)
+              }
+            case fail: TipUpdateResult.Failure =>
+              logger.warn(
+                s"Could not verify header=${header.hashBE.hex}, reason=$fail")
+              ConnectTipResult.BadTip(fail)
+          }
       }
-      parseSuccessOrFailure(nested)
     }
-
     tipResult
   }
 
-  /** Takes in a vector of blockchain updates being executed asynchronously, we can only connect one [[BlockHeader header]]
-    * to a tip successfully, which means _all_ other [[BlockchainUpdate updates]] must fail. This is a helper method
-    * to find the one [[BlockchainUpdate.Successful successful]] update, or else returns one of the [[BlockchainUpdate.Failed failures]]
-    * @return
-    */
-  private def parseSuccessOrFailure(
-      updates: Vector[BlockchainUpdate]): BlockchainUpdate = {
-    require(updates.nonEmpty,
-            s"Cannot parse success or failure if we don't have any updates!")
-    val successfulTipOpt: Option[BlockchainUpdate] = {
-      updates.find {
-        case update: BlockchainUpdate =>
-          update.isInstanceOf[BlockchainUpdate.Successful]
+  /** Iterates through each given blockchains attempting to connect the given headers to that chain
+    * @return The final updates for each chain
+    *
+    * */
+  def connectHeadersToChains(
+      headers: Vector[BlockHeader],
+      blockchains: Vector[Blockchain])(
+      implicit chainAppConfig: ChainAppConfig): Vector[BlockchainUpdate] = {
+    logger.debug(
+      s"Attempting to connect ${headers.length} headers to ${blockchains.length} blockchains")
+
+    @tailrec
+    def loop(
+        headersToProcess: List[BlockHeader],
+        lastUpdates: Vector[BlockchainUpdate]): Vector[BlockchainUpdate] = {
+      headersToProcess match {
+        case h :: t =>
+          val newUpdates: Vector[BlockchainUpdate] = lastUpdates
+            .map { lastUpdate =>
+              val connectTipResult =
+                Blockchain.connectTip(header = h,
+                                      blockchain = lastUpdate.blockchain)
+              parseConnectTipResult(connectTipResult, lastUpdate)
+            }
+            .flatten
+            .toVector
```

Comment:
I know I've been bringing this up many times, but feel like this is a good example of my point: We're doing a lot of list conversions at places we don't necessarily need to do so. Like going back and forth with `.toVector` here and `.toList` on line 165. I'm not sure about the performance characteristics of collections conversions, but I'm assuming it's not O(1). 

The arguments to `loop` could be `Seq[BlockHeader]`, and if we change line 144 to `case h +: t =>` the match would still work. Sadly, using `::` to match on `Vector`s don't work:-(

In general, I think [Postels law](https://en.wikipedia.org/wiki/Robustness_principle) has something going for it: we can accept `Seq[T]` as arguments and return `Vector[T]`